documentation/src/main/asciidoc/topics/con_performance_data_consistency.adoc

```diff
@@ -0,0 +1,105 @@
+
+Transactions are a sequence of atomic operations that {brandname} caries out as a single operation.
+Either all reads and writes in the transaction complete successfully or they all fail.
+In this way, the transaction either modifies the cache state in a consistent way, providing a history of reads and writes, or it does not modify cache state at all.
+
+The following is a simple transaction with one read and one write operation:
+
+[source,java,options="nowrap",subs=attributes+]
+----
+transaction.begin()
+cache.get(k1) // reads v1
+cache.put(k1, v2) // writes v2
+transaction.commit()
+----
+
+Because each one is composed of a series of operations, transactions increase the contention of read and write operations compared to non-transactional caches.
+
+The main performance concern for enabling transactions is finding the balance between having a more consistent data set and increasing latency that degrades write throughput.
+
+.Write locking
+
+To perform write operations while carrying out transactions, {brandname} locks keys.
+Locking keys adds contention that increases latency for write operations.
+How much contention depends on whether {brandname} locks keys optimistically or pessimistically.
+
+* Optimistic locking
++
+For low rates of contention, {brandname} locks keys when transactions invoke the `commit()` method.
+Keys are locked for shorter periods of time, which means less overall latency for write operations, but transaction failures are more costly.
+
+* Pessimistic locking
++
+For high rates of contention, {brandname} locks keys when transactions invoke the `put()` method.
+Keys are locked for longer periods of time, which increases latency for write operations, but makes transactions more efficient.
+
+Because they take longer to complete than performing individual operations, transaction failures are more costly in terms of excessive resource consumption.
+For example if a transaction takes five milliseconds to complete, with optimistic locking, you lose the entire five milliseconds.
+With pessimistic locking, if {brandname} detects a conflict after two milliseconds and the transaction rolls back, you lose only those two milliseconds instead.
+
+From that perspective having more efficient transactions with pessimistic locking might offset the reduction in overall write throughput.
+
+.Read isolation
+
+Isolation levels guarantee whether or not data in the cache has changed during transactions.
+
+.Isolation levels
+[%autowidth,cols="1,1",stripes=even]
+|===
+|`REPEATABLE_READ`
+|Read operations can only ever have one value in the context of a transaction. Once it retrieves the value of an entry, {brandname} returns that same value for any subsequent read operations.
+
+|`READ_COMMITTED`
+|Read operations can return different values if another transaction modifies keys. Each read operation in a transaction returns the most recent value for entries.
```

Comment:
"Each read operation in a transaction returns the most recent value for entries."
actually, ^ is not correct every time. If the node is not a read-owner of the entry, Infinispan will not fetch the most up to date value (otherwise, it would incur in a remote get)

Reply:
Changed to "Each read operation in a transaction returns the most recent value for entries as long as the node is a read-owner for the entry." Should we maybe drop this sentence though? Do we want to spell out what happens if the node is not a read-owner?

Reply:
maybe: "Read operations may return different values if another transaction modifies" :)

Reply:
@pruivo Thanks, I made some changes. Hopefully didn't butcher the meaning even further. :)